src/test/validation_chainstatemanager_tests.cpp

```diff
@@ -15,15 +18,62 @@
 
 #include <boost/test/unit_test.hpp>
 
-BOOST_FIXTURE_TEST_SUITE(validation_chainstatemanager_tests, TestingSetup)
+struct ChainTestingSetup : public BasicTestingSetup {
+    boost::thread_group threadGroup;
+
+    explicit ChainTestingSetup(const std::string& chainName = CBaseChainParams::MAIN, const std::vector<const char*>& extra_args = {})
+        : BasicTestingSetup(chainName, extra_args)
+    {
+        // Ideally we'd move all the RPC tests to the functional testing framework
+        // instead of unit tests, but for now we need these here.
+        RegisterAllCoreRPCCommands(tableRPC);
+
+        m_node.scheduler = MakeUnique<CScheduler>();
+
+        // We have to run a scheduler thread to prevent ActivateBestChain
+        // from blocking due to queue overrun.
+        threadGroup.create_thread([&] { TraceThread("scheduler", [&] { m_node.scheduler->serviceQueue(); }); });
+        GetMainSignals().RegisterBackgroundSignalScheduler(*m_node.scheduler);
+
+        pblocktree.reset(new CBlockTreeDB(1 << 20, true));
+
+        m_node.mempool = MakeUnique<CTxMemPool>(&::feeEstimator);
+
+        m_node.chainman = &::g_chainman;
+
+        // Start script-checking threads. Set g_parallel_script_checks to true so they are used.
+        constexpr int script_check_threads = 2;
+        for (int i = 0; i < script_check_threads; ++i) {
+            threadGroup.create_thread([i]() { return ThreadScriptCheck(i); });
+        }
+        g_parallel_script_checks = true;
+    }
+    ~ChainTestingSetup() {
+        if (m_node.scheduler) m_node.scheduler->stop();
+        threadGroup.interrupt_all();
+        threadGroup.join_all();
+        GetMainSignals().FlushBackgroundCallbacks();
+        GetMainSignals().UnregisterBackgroundSignalScheduler();
+        m_node.args = nullptr;
```

Comment:
Remove this line.

Reply:
I copied this from `~TestingSetup`, is there a reason why this is needed there and not here?

Reply:
It's a bit odd to have the low level code copied here, and require modifying `src/test/validation_chainstatemanager_tests.cpp` whenever the testing setup is upgraded.

wouldn't it be possible to simply keep using the `TestingSetup` here, like the tests did previously, and then do specific undo-hacks (hopefully less code) to unload the chainstates (and/or whatever else is required)

Reply:
oops. My mistake. This should be left. I was confusing it with the RPC registration/deregistration.

Reply:
@MarcoFalke I assume you mean something like:

```diff
diff --git a/src/test/validation_chainstatemanager_tests.cpp b/src/test/validation_chainstatemanager_tests.cpp
index 34e5723964..664405fc5e 100644
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -17,48 +17,7 @@
 
 #include <boost/test/unit_test.hpp>
 
-struct ChainTestingSetup : public BasicTestingSetup {
-    boost::thread_group threadGroup;
-
-    explicit ChainTestingSetup()
-        : BasicTestingSetup()
-    {
-        // We have to run a scheduler thread to prevent ActivateBestChain
-        // from blocking due to queue overrun.
-        m_node.scheduler = MakeUnique<CScheduler>();
-        threadGroup.create_thread([&] { TraceThread("scheduler", [&] { m_node.scheduler->serviceQueue(); }); });
-        GetMainSignals().RegisterBackgroundSignalScheduler(*m_node.scheduler);
-
-        pblocktree.reset(new CBlockTreeDB(1 << 20, true));
-
-        m_node.mempool = MakeUnique<CTxMemPool>(&::feeEstimator);
-
-        m_node.chainman = &::g_chainman;
-
-        // Start script-checking threads. Set g_parallel_script_checks to true so they are used.
-        constexpr int script_check_threads = 2;
-        for (int i = 0; i < script_check_threads; ++i) {
-            threadGroup.create_thread([i]() { return ThreadScriptCheck(i); });
-        }
-        g_parallel_script_checks = true;
-    }
-    ~ChainTestingSetup() {
-        if (m_node.scheduler) m_node.scheduler->stop();
-        threadGroup.interrupt_all();
-        threadGroup.join_all();
-        GetMainSignals().FlushBackgroundCallbacks();
-        GetMainSignals().UnregisterBackgroundSignalScheduler();
-        m_node.args = nullptr;
-        UnloadBlockIndex(m_node.mempool.get(), *m_node.chainman);
-        m_node.mempool.reset();
-        m_node.scheduler.reset();
-        m_node.chainman->Reset();
-        m_node.chainman = nullptr;
-        pblocktree.reset();
-    }
-};
-
-BOOST_FIXTURE_TEST_SUITE(validation_chainstatemanager_tests, ChainTestingSetup)
+BOOST_FIXTURE_TEST_SUITE(validation_chainstatemanager_tests, TestingSetup)
 
 //! Basic tests for ChainstateManager.
 //!
@@ -68,6 +27,9 @@ BOOST_AUTO_TEST_CASE(chainstatemanager)
     ChainstateManager& manager = *m_node.chainman;
     CTxMemPool& mempool = *m_node.mempool;
 
+    UnloadBlockIndex(nullptr, manager);
+    manager.Reset();
+
     std::vector<CChainState*> chainstates;
     const CChainParams& chainparams = Params();
 
@@ -151,6 +113,9 @@ BOOST_AUTO_TEST_CASE(chainstatemanager_rebalance_caches)
     ChainstateManager& manager = *m_node.chainman;
     CTxMemPool& mempool = *m_node.mempool;
 
+    UnloadBlockIndex(nullptr, manager);
+    manager.Reset();
+
     size_t max_cache = 10000;
     manager.m_total_coinsdb_cache = max_cache;
     manager.m_total_coinstip_cache = max_cache;
```

I found it hard to convince myself that `UnloadBlockIndex` and `Reset` cleanly resets everything exactly to before `ActivateBestChain` is run, which is why I thought repetition wasn't such a bad thing here. Let me know if I'm missing anything though.